### ISSUE TITLE
⚙️ Configure Dependabot For Terraform Modules CI/CD Dependencies

### ISSUE THREAD

**Issue by connormaglynn** [ministryofjustice/modernisation-platform#8211]
### User Story

As a Modernisation Platform Engineer
I need/want/expect to take minimal dependency version bumps when I release a new version of a Terraform Module, after I have made a functional change
So that I can reduce the time needed to debug any issues and easily identify the cause of any issues, whether that be a dependency version bump or the functional change

### Value / Purpose

The purpose of this work is to implement a new dependency management strategy for Terraform Modules, especially regarding CI/CD related dependencies (which tend to be the noisiest), to reduce the number of dependency related changes an engineer needs to release when they make a functional change; and to reduce the number of depency bumps come through in support.

The value of implementing this new strategy will be:
- Easier to release functional changes to Terraform Modules, less noise, less risk, easier to debug issues etc.
- Decrease support burden by reducing the number of Dependabot PRs needing to be merged

### Useful Contacts

@connormaglynn

### Additional Information

- This was discussed in the [Ways of Engineering](https://miro.com/app/board/uXjVPFK31Mw=/?moveToWidget=3458764601116040392&cot=14), which has some additional information
- Team members tended to like the idea of only performing releases when we make functional changes - so the new strategy should not be to release when we merge Dependabot PRs as these are mostly related to CI/CD
- Team members tended to dislike the volume of CI/CD related Dependabot commits - especially in low volume of change Terraform Modules where the CI/CD process isn't being used often
- [Here](https://docs.google.com/spreadsheets/d/10mYvzclUA49PYwmKzYlRCLw_WY1kDaQI2kUKnqTimH4/edit?usp=sharing) is a summary of our Terraform Modules, their last release date and how many commits have been made since the last release

#### Perscriptive Dependency Management Strategy

Here's a perspective on dependency management strategy we can use across all of our estates to reduce maintenance burdens, while still ensuring a high level of security and managing technical debt.

**Goal:** Reduce Dependabot PR noise while still patching security issues quickly and keeping technical debt under control.

- **Security vulnerabilities (highest priority):** Enable Dependabot alerts and security update PRs so vulnerable dependencies are detected and remediated quickly.
- **Major version upgrades (medium priority):** Raise grouped major-version update PRs every two weeks so we proactively address technical debt and any breaking changes.
- **Minor/patch upgrades (low priority):** Apply minor/patch updates every 6 months, drastically reducing maintenance burdens of constant dependabot churn. We should upgrade minor and patch versions if we need a new feature in a later version, but we don't always need to be on the latest minor or patch version.

**References**
- Dependabot alerts: https://docs.github.com/en/code-security/dependabot/dependabot-alerts/about-dependabot-alerts
- Dependabot security updates: https://docs.github.com/en/code-security/dependabot/dependabot-security-updates/about-dependabot-security-updates
- Dependabot groups: https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/optimizing-pr-creation-and-approval-with-dependabot-groups
- Dependabot version updates config: https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates
- tfsec: https://github.com/aquasecurity/tfsec
- Checkov: https://www.checkov.io/

### Definition of Done

- [ ] A new Dependency management strategy has been implemented across our Terraform Modules - which reduces the number of CI/CD related Dependabot PRs
- [ ] Ways of Engineering updated to reflect dependency management strategy

**Comment by connormaglynn** [comment]
I also wrote a bash script to go through our Terraform Modules and gather the last release date and commits since the last release - may be of use in the future 🔮🚀👇

```shell
#!/bin/bash

# Replace with your GitHub token
GITHUB_TOKEN=$(gh auth token)

# List of repository names (format: owner/repo)
repos=(
  "ministryofjustice/modernisation-platform-terraform-bastion-linux"
  "ministryofjustice/modernisation-platform-terraform-ecs-cluster"
  "ministryofjustice/modernisation-platform-terraform-s3-bucket"
  "ministryofjustice/modernisation-platform-terraform-aws-vm-import"
  "ministryofjustice/modernisation-platform-terraform-pagerduty-integration"
  "ministryofjustice/modernisation-platform-terraform-loadbalancer"
  "ministryofjustice/modernisation-platform-terraform-ssm-patching"
  "ministryofjustice/modernisation-platform-terraform-ec2-instance"
  "ministryofjustice/modernisation-platform-terraform-ec2-autoscaling-group"
  "ministryofjustice/modernisation-platform-terraform-lambda-function"
  "ministryofjustice/modernisation-platform-terraform-baselines"
  "ministryofjustice/modernisation-platform-terraform-cross-account-access"
  "ministryofjustice/modernisation-platform-terraform-environments"
  "ministryofjustice/modernisation-platform-terraform-iam-superadmins"
  "ministryofjustice/modernisation-platform-terraform-member-vpc"
  "ministryofjustice/modernisation-platform-github-oidc-provider"
)

# GitHub API base URL
api_url="https://api.github.com/repos"

# Loop through each repository
for repo in "${repos[@]}"; do
  # Fetch the latest release information using GitHub API with authentication
  release_response=$(curl -s -H "Authorization: token $GITHUB_TOKEN" "$api_url/$repo/releases/latest")

  # Extract the release date and the tag name of the latest release
  last_release_date=$(echo "$release_response" | jq -r '.published_at')
  last_release_tag=$(echo "$release_response" | jq -r '.tag_name')

  # Check if there was a valid release
  if [ "$last_release_date" != "null" ] && [ "$last_release_tag" != "null" ]; then
    # Get the commit SHA for the release tag
    last_release_sha=$(curl -s -H "Authorization: token $GITHUB_TOKEN" "$api_url/$repo/git/ref/tags/$last_release_tag" | jq -r '.object.sha')

    # Get the default branch (usually main/master)
    default_branch=$(curl -s -H "Authorization: token $GITHUB_TOKEN" "$api_url/$repo" | jq -r '.default_branch')

    # Get the latest commit SHA on the default branch
    latest_commit_sha=$(curl -s -H "Authorization: token $GITHUB_TOKEN" "$api_url/$repo/commits/$default_branch" | jq -r '.sha')

    # Count the number of commits since the last release
    commit_count=$(curl -s -H "Authorization: token $GITHUB_TOKEN" "$api_url/$repo/compare/$last_release_sha...$latest_commit_sha" | jq -r '.ahead_by')

    # Output the repository name, last release date, and number of commits since the last release
    echo "$repo,$last_release_date,$commit_count"
  else
    echo "$repo No releases found"
  fi
done
```

**Comment by ep-93** [comment]
A — 

Drop github-actions and gomod from daily to weekly; keep terraform separate and move it to monthly.
Add Dependabot auto-merge for the action-dependencies and gomod-dependencies groups, gated on the PR passing your full check suite (the dependabot/fetch-metadata action + gh pr merge --auto, restricted to patch/minor). Those groups then merge themselves on green.

Explicitly do not auto-merge the terraform ecosystem — the aws ~> 6.0 constraint is the one dependency that actually reaches your consumers, so it stays a human review and rides with a deliberate release.

This already delivers the user story: functional changes release clean (main is kept continuously current by auto-merge, so there's nothing pending to bundle), and the Dependabot support load on this repo drops to near zero. It's maybe a day of work and it's per-repo.

B — The bigger fix, bigger, estate-wide. Everything in A, plus lift the inlined leaf jobs (code-scanning, scorecards, go-terratest, and the checkout-only bits of documentation) up into workflow_call reusable workflows in modernisation-platform-github-actions, which you already own and call. Each module's workflow collapses to a thin caller pinned to one version of the central repo. Now actions/checkout bumping is one Dependabot PR centrally instead of one per module across the whole estate — that's the cross-estate support burden gone, not just this repo's. It keeps SHA-pinning intact, so Scorecards is unaffected. The cost is coupling: a bad central change can break many modules at once, so the central workflows need proper semver and a pilot before you bump the estate.

**Comment by ep-93** [comment]
Discussed with Ewa, going with A

**Comment by ep-93** [comment]
Applied here - https://github.com/ministryofjustice/modernisation-platform-terraform-ssm-patching/pull/870

Need to deploy in a few weeks to other repos. Will get the PRs ready. And then edit the template.

**Comment by ep-93** [comment]
https://github.com/ministryofjustice/modernisation-platform-terraform-loadbalancer/pull/918
https://github.com/ministryofjustice/modernisation-platform-terraform-member-vpc/pull/883

**Comment by ep-93** [comment]
https://github.com/ministryofjustice/modernisation-platform-terraform-ssm-patching/pull/873
https://github.com/ministryofjustice/modernisation-platform-terraform-bastion-linux/pull/928
